include/deal.II/numerics/vector_tools_project.templates.h

```diff
@@ -213,7 +216,64 @@ namespace VectorTools
         LinearAlgebra::distributed::Vector<Number>>;
       MatrixType mass_matrix;
       mass_matrix.initialize(matrix_free);
-      mass_matrix.compute_diagonal();
+
+      // For most elements we want to use the lumped mass matrix as the
+      // preconditioner. However, this isn't going to work with a lot of
+      // elements, like almost everything not defined on hypercubes: in that
+      // case use the diagonal.
+      bool        use_lumped = false;
+      const auto &fes        = dof.get_fe_collection();
+      if (dof.get_triangulation().all_reference_cells_are_hyper_cube())
+        {
+          // A few hypercube elements will always work, so avoid the more
+          // expensive check if we have them
+          bool all_support_mass_lumping = true;
+          for (unsigned int i = 0; i < fes.size(); ++i)
+            if (dynamic_cast<const FE_Q<dim, spacedim> *>(&fes[i]) == nullptr &&
+                dynamic_cast<const FE_DGQ<dim, spacedim> *>(&fes[i]) == nullptr)
```

Comment:
I like this, as this will also return non-null pointers for `FE_DGQArbitraryNodes`, as the `FE_DGQArbitraryNodes(QGauss(degree+1))` the lumped mass matrix will be the ultimate approximation, namely the exact diagonal (on affine mesh cells).

Reply:
It also works with `FE_DGQLegendre` and `FE_DGQHermite` (though those are merely first-order accurate approximations of the mass matrix)